nilearn/connectivity/connectivity_matrices.py

```diff
@@ -0,0 +1,349 @@
+    """
+    return np.dot(eigenvectors * function(eigenvalues), eigenvectors.T)
+
+
+def _map_eigenvalues(function, symmetric):
+    """Matrix function, for real symmetric matrices. The function is applied
+    to the eigenvalues of symmetric.
+
+    Parameters
+    ----------
+    function : function numpy.ndarray -> numpy.ndarray
+        The transform to apply to the eigenvalues.
+
+    symmetric : numpy.ndarray, shape (n_features, n_features)
+        The input symmetric matrix.
+
+    Returns
+    -------
+    output : numpy.ndarray, shape (n_features, n_features)
+        The new symmetric matrix obtained after transforming the eigenvalues,
+        while keeping the same eigenvectors.
+
+    Note
+    ----
+    If input matrix is not real symmetric, no error is reported but result will
+    be wrong.
+    """
+    eigenvalues, eigenvectors = linalg.eigh(symmetric)
+    return _form_symmetric(function, eigenvalues, eigenvectors)
+
+
+def _geometric_mean(matrices, init=None, max_iter=10, tol=1e-7):
+    """Compute the geometric mean of symmetric positive definite matrices.
+
+    The geometric mean of n positive definite matrices
+    M_1, ..., M_n is the minimizer of the sum of squared distances from an
+    arbitrary matrix to each input matrix M_k
+
+    gmean(M_1, ..., M_n) = argmin_X sum_{k=1}^N dist(X, M_k)^2
+
+    where the used distance is related to matrices logarithm
+
+    dist(X, M_k) = ||log(X^{-1/2} M_k X^{-1/2)}||
+
+    In case of positive numbers, this mean is the usual geometric mean.
+
+    References
+    ----------
+    See Algorithm 3 of:
+        P. Thomas Fletcher, Sarang Joshi. Riemannian Geometry for the
+        Statistical Analysis of Diffusion Tensor Data. Signal Processing, 2007.
+
+    Parameters
+    ----------
+    matrices : list of numpy.ndarray, all of shape (n_features, n_features)
+        List of matrices whose geometric mean to compute. Raise an error if the
+        matrices are not all symmetric positive definite of the same shape.
+
+    init : numpy.ndarray, shape (n_features, n_features), optional
+        Initialization matrix, default to the arithmetic mean of matrices.
+        Raise an error if the matrix is not symmetric positive definite of the
+        same shape as the elements of matrices.
+
+    max_iter : int, optional
+        Maximal number of iterations.
+
+    tol : positive float or None, optional
+        The tolerance to declare convergence: if the gradient norm goes below
+        this value, the gradient descent is stopped. If None, no  check is
+        performed.
+
+    Returns
+    -------
+    gmean : numpy.ndarray, shape (n_features, n_features)
+        Geometric mean of the matrices.
+    """
+    # Shape and symmetry positive definiteness checks
+    n_features = matrices[0].shape[0]
+    for matrix in matrices:
+        _check_square(matrix)
+        if matrix.shape[0] != n_features:
+            raise ValueError("Matrices are not of the same shape.")
```

Comment:
Could you please be more verbose? You could print the metrix shape and the expected one.